File: python/questions/getMinimumDifference.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        ans = float('inf')
        pre = None
        def dfs(node: Optional[TreeNode]) -> void:
            nonlocal pre, ans
            if node is None:
                return
            if node.left:
                dfs(node.left)
            ans = min(ans, node.val-pre.val) if pre else ans
            pre = node
            if node.right:
                dfs(node.right)
        dfs(root)
        return ans

File: cpp/questions/getMinimumDifference.cpp
#include <vector>
#include <climits>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

using namespace std;

class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        int ans = INT_MAX;
        TreeNode* pre = nullptr;
        auto dfs = [&](this auto&& dfs, TreeNode* node) -> void {
            if (node == nullptr) {
                return;
            }
            dfs(node->left);
            ans = pre ? std::min(ans, node->val-pre->val) : ans;
            pre = node;
            dfs(node->right);
        };
        dfs(root);
        return ans;
    }
};
